tools/nightly-e2e-tests/src/main_tf_renderer.py

```diff
@@ -0,0 +1,133 @@
+# -----------------------------------------------------------
+# Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+# This product includes software developed at Datadog (https://www.datadoghq.com/).
+# Copyright 2019-2020 Datadog, Inc.
+# -----------------------------------------------------------
+
+import re
+from src.test_file_parser import MonitorConfiguration, MonitorVariable, MONITOR_TYPE_LOGS, MONITOR_TYPE_APM
+from src.linter import Linter, linter_context
+
+
+class MainTF:
+    """
+    Generates `main.tf` file.
+    """
+    def __init__(
+            self,
+            template_content: str,
+            logs_monitor_template: 'MonitorTemplate',
+            apm_monitor_template: 'MonitorTemplate'
+    ):
+        self.monitors: [MonitorConfiguration] = []
+        self.template_content = template_content
+        self.logs_monitor_template = logs_monitor_template
+        self.apm_monitor_template = apm_monitor_template
+
+    @staticmethod
+    def load_from_templates(
+            main_template_path: str,
+            logs_monitor_template_path: str,
+            apm_monitor_template_path: str
+    ):
+        with open(logs_monitor_template_path, 'r') as monitor_tf_src:
+            logs_monitor_template = MonitorTemplate(template_content=monitor_tf_src.read())
+
+        with open(apm_monitor_template_path, 'r') as monitor_tf_src:
+            apm_monitor_template = MonitorTemplate(template_content=monitor_tf_src.read())
+
+        with open(main_template_path, 'r') as main_tf_src:
+            return MainTF(
+                template_content=main_tf_src.read(),
+                logs_monitor_template=logs_monitor_template,
+                apm_monitor_template=apm_monitor_template
+            )
+
+    def render(self, monitors: [MonitorConfiguration]):
+        output = '# This file is auto-generated, do not edit it directly\n\n'
+        output += f'{self.template_content}\n'
+        output += '# Monitors:\n\n'
+        for monitor in monitors:
+            monitor_template: MonitorTemplate
+
+            if monitor.type == MONITOR_TYPE_LOGS:
+                monitor_template = self.logs_monitor_template
+            else:
+                assert monitor.type == MONITOR_TYPE_APM, f'Unrecognized monitor type {monitor.type}'
+                monitor_template = self.apm_monitor_template
+
+            output += monitor_template.render(monitor=monitor)
+            output += '\n'
+        return output
+
+
+class MonitorTemplate:
+    """
+    Renders `monitor-logs.tf.src` file by replacing variables with values read from `MonitorConfiguration`.
+    If a given `$variable` is not defined in monitor configuration, then its default from template will be used.
+    """
+    def __init__(self, template_content: str):
+        self.template_content = template_content
+
+    def render(self, monitor: MonitorConfiguration):
```

Comment:
are you missing the return type on all these functions or is this intentionally ?

Reply:
Indeed, it was missing in a few methods - I added return types in https://github.com/DataDog/dd-sdk-ios/pull/521/commits/3e16ca074809f31eb7e426c8bd36e1bca2136171. Python uses [Duck typing](https://en.wikipedia.org/wiki/Duck_typing) so it's type control is not that strict as Swift or Kotlin, thus Pycharm doesn't even warn you if you ommit the type declaration. This can be perhaps enforced with some additional plugins or configuration.